node/heartbeat/hearbeatMessageInfo.go

```diff
@@ -47,39 +54,54 @@ func (hbmi *heartbeatMessageInfo) clockTime() time.Time {
 	return time.Now()
 }
 
-// Sweep updates all records
-func (hbmi *heartbeatMessageInfo) sweep() {
+func (hbmi *heartbeatMessageInfo) updateFields() {
 	crtDuration := hbmi.getTimeHandler().Sub(hbmi.timeStamp)
+	crtDuration = protectAgainstNegativeDuration(crtDuration)
+
 	hbmi.isActive = crtDuration < hbmi.maxDurationPeerUnresponsive
 	hbmi.updateUpAndDownTime()
 	hbmi.updateMaxInactiveTimeDuration()
 }
 
 // Wil update the total time a node was up and down
 func (hbmi *heartbeatMessageInfo) updateUpAndDownTime() {
+	lastDuration := hbmi.clockTime().Sub(hbmi.lastTimeUptimeDowntime)
+	lastDuration = protectAgainstNegativeDuration(lastDuration)
+
 	if hbmi.isActive {
-		hbmi.totalUpTime.Duration += hbmi.clockTime().Sub(hbmi.timeStamp)
+		hbmi.totalUpTime.Duration += lastDuration
 	} else {
-		if hbmi.timeStamp != emptyTimestamp {
-			hbmi.totalDownTime.Duration += hbmi.clockTime().Sub(hbmi.timeStamp)
-		}
+		hbmi.totalDownTime.Duration += lastDuration
 	}
+
+	hbmi.lastTimeUptimeDowntime = time.Now()
 }
 
 // HeartbeatReceived processes a new message arrived from a peer
 func (hbmi *heartbeatMessageInfo) HeartbeatReceived(shardID uint32, version string, nodeDisplayName string) {
 	crtTime := hbmi.getTimeHandler()
-	hbmi.sweep()
+	hbmi.updateFields()
 	hbmi.shardID = shardID
 	hbmi.updateMaxInactiveTimeDuration()
 	hbmi.timeStamp = crtTime
 	hbmi.versionNumber = version
 	hbmi.nodeDisplayName = nodeDisplayName
+	hbmi.updateFields()
 }
 
 func (hbmi *heartbeatMessageInfo) updateMaxInactiveTimeDuration() {
 	crtDuration := hbmi.getTimeHandler().Sub(hbmi.timeStamp)
+	crtDuration = protectAgainstNegativeDuration(crtDuration)
```

Comment:
this is a minFunction. like MIN(0, value). you could do directly that from crtDuration := MIN(0, hbmi.getTimeHanlder().Sub(hbmi.timeStamp)

Reply:
done with max